### PR TITLE
Translate `_d_newitem{U,T,iT}` to a template

### PR DESCRIPTION
**Pull request by teodutu**
This contributes to https://github.com/dlang/projects/issues/25 and makes the following changes:
- Replace `_d_newitem{U,T,iT}` with a template `_d_newitemT` that allocates memory for the new `struct` and performs the required initialisation
- Move lowering call to `_d_newitemT` from e2ir.d to expressionsem.d
- Add `lowering` field to `NewExp` that contains the lowered call and use it in e2ir.d
- Remove `_d_newitem{,i}T` from rt.lifetime.d

### CODE REVIEW

**Comment by UplinkCoder** [comment]
tests seem to be failing.

**Comment by teodutu** [comment]
Yeah @UplinkCoder, `uniformctor.d` fails because `__traits(isZeroInit)` seems to be true for enums. I'm not sure why `implicit.d` fails yet.

File: druntime/test/profile/bothgc.log.exp [hunk 1]
```diff
@@ -1,2 +1,3 @@
 bytes allocated, allocations, type, function, file:line
-          16000	           1000	both.Num both.foo src/both.d:15
+          20496	           1000	Num both.foo src/both.d:15
+          16000	           1000	both.Num core.lifetime._d_newitemT!(Num)._d_newitemT ../../src/core/lifetime.d:2748
```
Comment (teodutu) [inline druntime/test/profile/bothgc.log.exp:3]: This kind of changes should not happen. They mean the new hook allocates more memory than the previous one, which should not be the case. I'll try to fix them.

File: druntime/test/profile/myprofilegc.log.linux.64.exp [hunk 2]
```diff
@@ -11,8 +11,10 @@ bytes allocated, allocations, type, function, file:line
              16	              1	char[] D main src/profilegc.d:36
              16	              1	closure profilegc.main.foo src/profilegc.d:45
              16	              1	float D main src/profilegc.d:16
+             16	              1	float core.lifetime._d_newitemT!float._d_newitemT ../../src/core/lifetime.d:2748
              16	              1	float[] D main src/profilegc.d:17
              16	              1	int D main src/profilegc.d:13
+             16	              1	int core.lifetime._d_newitemT!int._d_newitemT ../../src/core/lifetime.d:2748
```
Comment (teodutu) [inline druntime/test/profile/myprofilegc.log.linux.64.exp:17]: Ditto.

File: druntime/test/profile/bothgc.log.exp (hunk 1 above)
Comment (teodutu) [reply druntime/test/profile/bothgc.log.exp:3]: This was actually incorrect. I later found out this is caused by the call to `GC.qalloc()` within the template hook, which I cannot get past.

**Comment by teodutu** [comment]
`runnable/implicit.d` is failing because of this bug: https://issues.dlang.org/show_bug.cgi?id=23650.

**Comment by teodutu** [comment]
I managed to avoid https://issues.dlang.org/show_bug.cgi?id=23650 by avoiding the lowering altogether inside `__traits(compiles)`.

**Comment by PetarKirov** [comment]
@teodutu please rebase

**Comment by teodutu** [comment]
Done @PetarKirov. I'll look into the traceGC errors later this week.

File: compiler/src/dmd/expressionsem.d [hunk 3]
```diff
@@ -3574,6 +3574,46 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+    /**
+     * Sets the `lowering` field of a `NewExp` to a call to `_d_newitemT` unless
+     * compiling with `-betterC` or within `__traits(compiles)`.
+     *
+     * Params:
+     *  ne = the `NewExp` to lower
+     */
+    private void tryLowerToNewItem(NewExp ne)
+    {
+        if (global.params.betterC || sc.flags & SCOPE.compile)
```
Comment (RazvanN7) [inline compiler/src/dmd/expressionsem.d:3586]:
```suggestion
        if (global.params.betterC || sc.needsCodegen())
```

**Comment by RazvanN7** [comment]
cc @dkorpel @PetarKirov this seems to be green now.

**Review by RazvanN7 — APPROVED** [review]

File: druntime/src/core/lifetime.d [hunk 4]
```diff
@@ -2777,6 +2777,71 @@ if (is(T == class))
...
+        mixin(TraceHook!(T.stringof, "_d_newclassT"));
+
+        return _d_newclassT!T();
+    }
+    else
+        assert(0, "Cannot create new class if compiling without support for runtime type information!");
+}
+
+/**
+ * Allocate an initialised non-array item.
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2797]:
```suggestion
 * Allocate an initialised non-array item.
 *
```
So DDoc separates title and description

File: druntime/src/core/lifetime.d [hunk 5]
```diff
@@ -2805,15 +2870,135 @@ if (is(T == class))
...
+    auto s1 = new S1;
+    assert(s1.x == 5);
+    assert(GC.getAttr(s1) == GC.BlkAttr.NO_SCAN);
+
+    auto s2 = new S2(3);
+    assert(s2.x == 3);
+    assert(GC.getAttr(s2) == GC.BlkAttr.NO_SCAN);
+
+    auto s3 = new S3(1);
+    assert(s3.x == [1,1,1,1]);
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2940]:
```suggestion
    assert(s3.x == [1, 1, 1, 1]);
```

File: druntime/src/core/lifetime.d [hunk 6]
```diff
@@ -2805,15 +2870,135 @@ if (is(T == class))
...
+        {
+            E exc;
+            ~this() { throw exc; }
+        }
+
+        bool caught = false;
+        S1* s = new S1(new E("test onFinalizeError"));
+        try
+        {
+            GC.runFinalizers((cast(char*)(typeid(S1).xdtor))[0..1]);
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2969]:
```suggestion
            GC.runFinalizers((cast(char*)(typeid(S1).xdtor))[0 .. 1]);
```

File: druntime/src/core/lifetime.d [hunk 7]
```diff
@@ -2805,15 +2870,135 @@ if (is(T == class))
...
+version(D_ProfileGC)
+{
+    /**
+    * TraceGC wrapper around $(REF _d_newitemT, core,lifetime).
+    */
+    T* _d_newitemTTrace(T)(string file, int line, string funcname) @trusted
+    {
+        version (D_TypeInfo)
+        {
+            import core.internal.array.utils: TraceHook, gcStatsPure, accumulatePure;
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2996]:
```suggestion
            import core.internal.array.utils : TraceHook, gcStatsPure, accumulatePure;
```

File: druntime/src/rt/lifetime.d [hunk 8]
```diff
@@ -1148,25 +1148,8 @@ extern (C) void[] _d_newarraymiTX(const TypeInfo ti, size_t[] dims) @weak
...
-struct Si {int x = 3;}
-
-void main()
-{
-    new Sz(); // _d_newitemT(typeid(Sz))
-    new Si(); // _d_newitemiT(typeid(Si))
-}
----
+Non-template version of $(REF _d_newitemT, core,lifetime) that does not perform
+initialisation. Needed for $(REF allocEntry, rt,aaA).
```
Comment (dkorpel) [inline druntime/src/rt/lifetime.d:1152]:
```suggestion
initialization. Needed for $(REF allocEntry, rt,aaA).
```

File: druntime/src/core/lifetime.d [hunk 9]
```diff
@@ -2805,15 +2870,135 @@ if (is(T == class))
...
+}
+
+// Test initialisers
+@safe unittest
 {
-    version (D_TypeInfo)
     {
-        import core.internal.array.utils: TraceHook, gcStatsPure, accumulatePure;
-        mixin(TraceHook!(T.stringof, "_d_newclassT"));
+        // zero-initialisation
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2886]:
```suggestion
        // zero-initialization
```

File: druntime/src/core/lifetime.d [hunk 10]
```diff
@@ -2805,15 +2870,135 @@ if (is(T == class))
...
+// Test allocation
+@safe unittest
+{
+    struct S { }
+    S* s = _d_newitemT!S();
+
+    assert(s !is null);
+}
+
+// Test initialisers
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2882]:
```suggestion
// Test initializers
```

File: druntime/src/core/lifetime.d [hunk 11]
```diff
@@ -2777,6 +2777,71 @@ if (is(T == class))
...
+    if (tiSize)
+        flags |= GC.BlkAttr.STRUCTFINAL | GC.BlkAttr.FINALIZE;
+
+    auto blkInfo = GC.qalloc(totalSize, flags, null);
+    auto p = blkInfo.base;
+
+    if (tiSize)
+    {
+        // The GC might not have cleared the padding area in the block.
+        *cast(TypeInfo*)(p + (itemSize & ~(size_t.sizeof - 1))) = null;
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2836]:
```suggestion
        *cast(TypeInfo*) (p + (itemSize & ~(size_t.sizeof - 1))) = null;
```

File: druntime/src/core/lifetime.d [hunk 12]
```diff
@@ -2777,6 +2777,71 @@ if (is(T == class))
...
+        *cast(TypeInfo*)(p + blkInfo.size - tiSize) = cast() typeid(Unq);
```
Comment (dkorpel) [inline druntime/src/core/lifetime.d:2837]:
```suggestion
        *cast(TypeInfo*) (p + blkInfo.size - tiSize) = cast() typeid(Unq);
```

File: compiler/src/dmd/expressionsem.d [hunk 13]
```diff
@@ -3574,6 +3574,46 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+        id = new DotIdExp(ne.loc, id, Id.object);
+        auto tiargs = new Objects();
+        auto t = ne.type.nextOf.unqualify(MODFlags.wild);  // Remove `inout`
+        tiargs.push(t);
+        id = new DotTemplateInstanceExp(ne.loc, id, hook, tiargs);
+
+        auto arguments = new Expressions();
+        if (global.params.tracegc)
+        {
+            auto funcname = (sc.callsc && sc.callsc.func) ?
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:3606]: Is this branch tested? Codecov sees no coverage.

File: compiler/src/dmd/expressionsem.d [hunk 14]
```diff
@@ -3574,6 +3574,46 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+    private void tryLowerToNewItem(NewExp ne)
+    {
+        if (global.params.betterC || !sc.needsCodegen())
+            return;
+
+        auto hook = global.params.tracegc ? Id._d_newitemTTrace : Id._d_newitemT;
+        if (!verifyHookExist(ne.loc, *sc, hook, "new struct"))
+            return;
+
+        /* Lower the memory allocation and initialisation of `new T()` to
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:3593]:
```suggestion
        /* Lower the memory allocation and initialization of `new T()` to
```

**Review by dkorpel — CHANGES_REQUESTED** [review]
Mostly style nits

File: compiler/src/dmd/expressionsem.d (hunk 13 above)
Comment (teodutu) [reply compiler/src/dmd/expressionsem.d:3606]: Yes, this is how `_d_newitemTTrace()` is inserted. It's tested in druntime: https://github.com/dlang/dmd/blob/master/druntime/test/profile/src/profilegc.d.

File: druntime/src/core/lifetime.d [hunk 15]
```diff
@@ -2777,6 +2777,71 @@ if (is(T == class))
...
+ * Returns:
+ *     newly allocated item
+ */
+T* _d_newitemT(T)() @trusted
+{
+    import core.internal.lifetime : emplaceInitializer;
+    import core.internal.traits : hasElaborateDestructor, hasIndirections, Unqual;
+    import core.memory : GC;
+
+    alias Unq = Unqual!T;
```
[the remark this answers is not in the document]
Comment (teodutu) [reply druntime/src/core/lifetime.d:2821]: Good idea. Done.

**Comment by teodutu** [comment]
Done @dkorpel. Take another look when you can.

**Review by dkorpel — APPROVED** [review]
I thought you were going to forward `Unqual!T` in druntime, but unqualifying in the compiler works too.

**Pull request merged**